File: searching_ALGOs/binary_search.py
# in python we can directly use my_list.index(<item-to-be-searched>) to find index of an item in a list.

def binary_search(li,item):
    start=0
    end=len(li)-1
    while start<=end:
        mid=int((start+end)/2)
        if li[mid]==item:
            return mid
        elif li[mid]>item:
            end=mid-1
        else:
            start=mid+1
            
    return -1
    
my_list=[12,24,39,44,56]
index_of_item=binary_search(my_list,44)

File: searching_ALGOs/binary_search.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> vec;

#define INPUT(_x) int _x;cin>>_x;
#define FOR(i,a,b) for(int i=a;i<=b;i++)
#define TCASE int _t;cin>>_t;FOR(int i,1,_t)
#define VECTORFILL(_vec) int _x;cin>>_x;while(_x!=-1){vec.push_back(_x);cin>>_x;}
#define VECTORPRINT(_vec) int _t=0;while(_t<_vec.size()){cout<<_vec[_t++]<<endl;}

int binarySearch(int a,int start,int end){       //returns the index or -1 if not found
	if (start>end) return -1;
	int mid=(start+end)/2;
	if(vec[mid]==a) return mid;
	else if(vec[mid]>a) return binarySearch(a,start,mid-1);
	else if(vec[mid]<a) return binarySearch(a,mid+1,end);
}

int main()
{
	VECTORFILL(vec);      //assuming -1 is the breaking point
	INPUT(a);
	cout<<binarySearch(a,0,vec.size()-1);
	return 0;
}
